sql: make []Expr nodes into structs

RaduBerinde opened the pull request:
AST nodes Row, Tuple, Array, Values are slices that implement the Expr or
Statement interfaces. This has some unfortunate consequences:
- we cannot use == to check if an Expr/Statement has changed
- whenever we visit such a node (WalkExpr/WalStmt), we will always allocate a
  new copy of the node (the visitor always returns a newly allocated copy).

We change these types to structs and change the interface implementations to use
pointer receivers.

The value-based leaves (IntVal, DBool, etc) also have the latter problem; we
should consider changing them as well.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/4569)

<!-- Reviewable:end -->

tamird commented:
>  - we cannot use `==` to check if an Expr/Statement has changed

This new implementation still does not allow you to do this. Am I missing something?

> whenever we visit such a node (WalkExpr/WalkStmt), we will always allocate a
   new copy of the node (the visitor always returns a newly allocated copy).

Yes, that's true, but those are copies of a slice (which is copying 3 pointer-sized values), not a deep copy of the underlying array.

This change seems fine, but I'm not sure it gets us anything. Remember, changing to a pointer receiver means you're paying an extra indirection cost on every method call.

---

Reviewed 17 of 17 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4569#-:-KB9HLWR7sEn59iLlz6G:-612412390)*
<!-- Sent from Reviewable.io -->

petermattis commented:
LGTM

I don't think there will be any change in the number of allocations. Previously we were storing a slice into an interface which required an allocation. Now we're making that allocation explicit and storing the slice in a struct and a pointer to the struct in the interface.

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4569#-:-KB9NuaQBSFXZv5kQIwN:-1487594231)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
> This new implementation still does not allow you to do this. Am I missing something?
We can now compare two Exprs with `==` to see if a function like Walk or Visit returned the same node or a new one. See the "RFC for new WalkExpr" thread.

> Remember, changing to a pointer receiver means you're paying an extra indirection cost on every method call.
I'm not convinced there is any extra cost (given that in our case the receiver is on the heap regardless). We can talk more offline if you want.

---







*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4569#-:-KB9ggfTnSOmsHMJWfz0:1036169788)*
<!-- Sent from Reviewable.io -->

RaduBerinde committed:
sql: make []Expr nodes into structs

AST nodes Row, Tuple, Array, Values are slices that implement the Expr or
Statement interfaces. This has some unfortunate consequences:
 - we cannot use `==` to check if an Expr/Statement has changed
 - whenever we visit such a node (WalkExpr/WalkStmt), we will always allocate a
   new copy of the node (the visitor always returns a newly allocated copy).

We change these types to structs and change the interface implementations to use
pointer receivers.

The value-based leaves (IntVal, DBool, etc) also have the latter problem; we
should consider changing them as well.

tamird commented:
Oh, because you're comparing pointers. I see. This doesn't require you to have a struct; you could have just changed the definitions to `type Datum *[]Expr`. I'm pointing this out just for posterity; I agree that structs are nicer.

Ah, this is doing dynamic dispatch. Yes, I agree that there's not any extra cost.

LGTM

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4569#-:-KB9iQN7uZWQBpcOQgqR:-1400349322)*
<!-- Sent from Reviewable.io -->
